iroha/src/isi.rs

```diff
@@ -1,94 +1,109 @@
-use crate::{account, asset, domain, peer, prelude::*, wsv::WorldStateView};
+use crate::prelude::*;
 use iroha_derive::Io;
 use parity_scale_codec::{Decode, Encode};
 
 pub mod prelude {
     //! Re-exports important traits and types. Meant to be glob imported when using `Iroha`.
-    pub use crate::{account::isi::*, asset::isi::*, domain::isi::*, peer::isi::*};
+    pub use crate::{account::isi::*, asset::isi::*, domain::isi::*, isi::*, peer::isi::*};
 }
 
-/// Iroha provides a library of smart contracts called **I**roha **S**pecial **I**nstructions (ISI).
-/// To execute logic on the ledger, these smart contracts can be invoked via either transactions
-/// or registered event listeners.
-/// This trait represents API which every ISI should be aligned with.
-pub trait Instruction {
-    /// To execute the instruction this method implementation supplied with a mutable reference
-    /// to `WorldStateView`. It's responsibility of the instruction to keep `WSV` in a consistent
-    /// state and return `Err` in case of errors.
-    fn execute(&self, world_state_view: &mut WorldStateView) -> Result<(), String>;
+#[derive(Clone, Debug, Io, Encode, Decode)]
+pub enum Instruction {
+    Peer(crate::peer::isi::PeerInstruction),
+    Domain(crate::domain::isi::DomainInstruction),
+    Asset(crate::asset::isi::AssetInstruction),
+    Account(crate::account::isi::AccountInstruction),
+    Compose(Box<Instruction>, Box<Instruction>),
 }
 
-///
-#[derive(Clone, Debug, PartialEq, Io, Encode, Decode)]
-pub enum Contract {
-    AddSignatory(account::isi::AddSignatory),
-    AppendRole(account::isi::AppendRole),
-    CreateAccount(account::isi::CreateAccount),
-    CreateRole(account::isi::CreateRole),
-    AddAssetQuantity(asset::isi::AddAssetQuantity),
-    TransferAsset(asset::isi::TransferAsset),
-    CreateAsset(asset::isi::CreateAsset),
-    CreateDomain(domain::isi::CreateDomain),
-    AddPeer(peer::isi::AddPeer),
+impl Instruction {
```

Comment:
Philosophically, the new way of doing instructions is nice. However, in implementation this code has gone from being completely understandable by anyone to being completely obtuse. Just looking at this, you have no idea what is going on and have to go through multiple files and probably read English prose to figure it out. Whereas AddPeer or TransferAsset are self documenting and unmistakable.

Reply:
I'm not really sure I am convinced that the current approach is really better than what we started with. You still have to write rust code and compile it to add more core functionality. If you want composability like with blockly, that is a UI issue and doesn't change this backend work.

Reply:
https://soramitsu.atlassian.net/browse/RH2-93